onnxruntime/core/providers/cuda/tensor/concat_impl.cu

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/cuda/cuda_common.h"
+#include "concat_impl.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T>
+__global__ void _ConcatKernel(const fast_divmod block_size_including_axis_dim_div,
+                              const fast_divmod block_size_inside_axis_dim_div,
+                              const int64_t* concat_sizes,
+                              const int64_t* concat_sizes_range,
+                              const int num_inputs,
+                              T* output_data,
+                              const void** input_ptr,
+                              const CUDA_LONG N) {
+  CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
+  CUDA_LONG input_pos = 0;
+
+  int outter_block_index = 0;
+  int block_index = 0;
+  int offset = 0;
+
+  int input_index = 0;
+  int block_offset = 0;
+
+  block_size_including_axis_dim_div.divmod(id, outter_block_index, offset);
+  block_size_inside_axis_dim_div.divmod(offset, block_index, offset);
+
+  for (int i = 0; i < num_inputs; ++i) {
+    int64_t range_left = (i == 0) ? 0 : concat_sizes_range[i - 1];
+    if ((range_left <= block_index) && (block_index < concat_sizes_range[i])) {
+      input_index = i;
+      block_offset = block_index - range_left;
+      break;
+    }
+  }
```

Comment:
If you want to optimize for hundreds of inputs, running this loop for every CUDA thread might still be very expensive. I think it might be better to set it up as a lookup table in CPU and reuse, so CUDA only need to deal with a simple lookup.

Reply:
will update it, same for split

---
In reply to: [299692803](https://github.com/microsoft/onnxruntime/pull/1333#discussion_r299692803) [](ancestors = 299692803)